src/config.ts

```diff
@@ -1,103 +1,155 @@
  */
-function isConfig(c: ConfigFile): c is Config {
-  return typeof (c as Config).monorepo === 'object';
-}
+export default class Config {
+  /**
+   * @protected Is constructed by Config.read(configFile), try Config.getAll() instead
+   */
+  protected constructor(
+    public readonly file: ConfigFile,
+    public readonly monorepo: MonorepoConfig,
+    public steps: Step[],
+    public readonly env: Record<string, string>
+  ) {}
+
+  /**
+   * Base build we're comparing against, if one can be found. If one can't, we'll enter fallback mode and run
+   * everything.
+   */
+  public buildId?: string;
+
+  /**
+   * Set of changes that match the configuration
+   */
+  public changes: string[] = [];
+
+  /**
+   * Whether this config is currently considered for inclusion in the final pipeline output
+   */
+  public included?: boolean = false;
+
+  /**
+   * Reason for inclusion/exclusion
+   *
+   * Should fit into the sentence: "Foo has been included/excluded because it has {REASON}"
+   */
+  public reason = 'no matching changes';
+
+  public decide(included: boolean, reason: string): void {
+    this.included = included;
+    this.reason = reason;
+  }
 
-/**
- * Searches for all pipeline files anywhere in the current working directory, and returns their file locations
- */
-async function getConfigFiles(): Promise<ConfigFile[]> {
-  const cwd = process.cwd();
-
-  try {
-    const pipelines = await glob(PIPELINE_FILE_GLOB, {
-      dot: true,
-      cwd,
-      ignore: PIPELINE_FILE_IGNORE_GLOB,
-    });
-
-    return pipelines.map((path) => {
-      return {
-        basePath: cwd,
-        path,
-      };
-    });
-  } catch (e) {
-    log(e);
-    return [];
+  public mapSteps(mapFn: (s: Step) => Step): void {
+    this.steps = this.steps.map(mapFn);
   }
-}
 
-function strings(v: undefined | string[] | string): string[] {
-  if (!v || v.length <= 0) {
-    return [];
+  public useFallback(): void {
+    this.changes = ['fallback'];
+    this.buildId = undefined;
   }
-  if (_.isArray(v)) {
-    return v;
+
+  public setBuildId(buildId: string): void {
+    this.buildId = buildId;
   }
-  return [String(v)];
-}
 
-function nameFromFilename(config: ConfigFile): string | undefined {
-  const match = PIPELINE_FILE.exec(basename(config.path));
-  return match?.groups?.name || undefined;
-}
+  public async getContentHash(hasher: FileHasher): Promise<string> {
+    return hasher.hashMany(await this.getMatchingFiles());
+  }
+
+  /**
+   * Used as a key in global namespaces
+   */
+  public getComponent(): string {
+    const { pipeline } = getBuildkiteInfo();
+    return `${pipeline}/${this.monorepo.name}`;
+  }
+
+  /**
+   * Returns all files that match the pipeline, wether they have changes or not
+   */
+  public async getMatchingFiles(): Promise<string[]> {
+    const cwd = process.cwd();
+
+    const files: Promise<string[]> = Promise.all(
```

Comment:
This should include `this.file.path` as we want changes to the pipeline file to bust the cache.